### PR TITLE
added WithGivenObjects-objects variants

### PR DESCRIPTION
**Pull request by mohamed-barakat**
* IsomorphismFromDualToInternalHomWithGivenObjects
* IsomorphismFromInternalCoHomToCoDualWithGivenObjects

### CODE REVIEW

File: MonoidalCategories/gap/SymmetricClosedMonoidalCategoriesDerivedMethods.gi [hunk 1]
```diff
@@ -907,3 +907,27 @@ AddDerivationToCAP( MonoidalPostComposeMorphismWithGivenObjects,
     
 end : CategoryFilter := IsSymmetricClosedMonoidalCategory and IsStrictMonoidalCategory,
       Description := "MonoidalPostComposeMorphismWithGivenObjects using evaluation, and tensor hom adjunction" );
+
+##
+AddDerivationToCAP( IsomorphismFromDualToInternalHom,
+        
+  function( cat, a )
```
Comment (TKuh) [inline MonoidalCategories/gap/SymmetricClosedMonoidalCategoriesDerivedMethods.gi:914]: I would prefer a diagram for easier reading. That would also be consistent with the other derivations.

File: MonoidalCategories/gap/SymmetricClosedMonoidalCategoriesDerivedMethods.gi [hunk 2]
```diff
@@ -907,3 +907,27 @@ AddDerivationToCAP( MonoidalPostComposeMorphismWithGivenObjects,
...
+                   MorphismFromTensorProductToInternalHom( cat, a, TensorUnit( cat ) ) );
+    
+end : CategoryFilter := IsSymmetricClosedMonoidalCategory,
+      Description := "IsomorphismFromDualToInternalHom using MorphismFromTensorProductToInternalHom" );
+
+##
+AddDerivationToCAP( IsomorphismFromDualToInternalHomWithGivenObjects,
+        
+  function( cat, source, a, range )
+
```
Comment (TKuh) [inline MonoidalCategories/gap/SymmetricClosedMonoidalCategoriesDerivedMethods.gi:927]: Diagram.

File: MonoidalCategories/gap/SymmetricClosedMonoidalCategoriesDerivedMethods.gi [hunk 3]
```diff
@@ -907,3 +907,27 @@ AddDerivationToCAP( MonoidalPostComposeMorphismWithGivenObjects,
...
+      Description := "IsomorphismFromDualToInternalHom using MorphismFromTensorProductToInternalHom" );
+
+##
+AddDerivationToCAP( IsomorphismFromDualToInternalHomWithGivenObjects,
+        
+  function( cat, source, a, range )
+    
+    return PreCompose( cat,
+                   RightUnitorInverseWithGivenTensorProduct( cat, a, source ),
+                   MorphismFromTensorProductToInternalHomWithGivenObjects( cat, source, a, TensorUnit( cat ), range ) );
```
Comment (TKuh) [inline MonoidalCategories/gap/SymmetricClosedMonoidalCategoriesDerivedMethods.gi:930]: We have `source = a^v` but `MorphismFromTensorProductToInternalHomWithGivenObjects` expects the second argument to be `a^v x 1`.

File: MonoidalCategories/gap/SymmetricCoclosedMonoidalCategoriesDerivedMethods.gi [hunk 4]
```diff
@@ -919,3 +919,27 @@ AddDerivationToCAP( MonoidalPostCoComposeMorphismWithGivenObjects,
     
 end : CategoryFilter := IsSymmetricCoclosedMonoidalCategory and IsStrictMonoidalCategory,
       Description := "MonoidalPostCoComposeMorphismWithGivenObjects using coclosed evaluation, and cohom tensor adjunction" );
+
+##
+AddDerivationToCAP( IsomorphismFromInternalCoHomToCoDual,
+        
+  function( cat, a )
+
```
Comment (TKuh) [inline MonoidalCategories/gap/SymmetricCoclosedMonoidalCategoriesDerivedMethods.gi:927]: Diagram.

File: MonoidalCategories/gap/SymmetricCoclosedMonoidalCategoriesDerivedMethods.gi [hunk 5]
```diff
@@ -919,3 +919,27 @@ AddDerivationToCAP( MonoidalPostCoComposeMorphismWithGivenObjects,
...
+                   RightUnitorWithGivenTensorProduct( cat, a ) );
+    
+end : CategoryFilter := IsSymmetricCoclosedMonoidalCategory,
+      Description := "IsomorphismFromInternalCoHomToCoDual using MorphismFromInternalCoHomToTensorProduct" );
+
+##
+AddDerivationToCAP( IsomorphismFromInternalCoHomToCoDualWithGivenObjects,
+        
+  function( cat, source, a, range )
+
```
Comment (TKuh) [inline MonoidalCategories/gap/SymmetricCoclosedMonoidalCategoriesDerivedMethods.gi:939]: Diagram.

File: MonoidalCategories/gap/SymmetricCoclosedMonoidalCategoriesDerivedMethods.gi [hunk 6]
```diff
@@ -919,3 +919,27 @@ AddDerivationToCAP( MonoidalPostCoComposeMorphismWithGivenObjects,
...
+      Description := "IsomorphismFromInternalCoHomToCoDual using MorphismFromInternalCoHomToTensorProduct" );
+
+##
+AddDerivationToCAP( IsomorphismFromInternalCoHomToCoDualWithGivenObjects,
+        
+  function( cat, source, a, range )
+    
+    return PreCompose( cat,
+                   MorphismFromInternalCoHomToTensorProductWithGivenObjects( cat, source, TensorUnit( cat ), a, range ),
+                   RightUnitorWithGivenTensorProduct( cat, a, range ) );
```
Comment (TKuh) [inline MonoidalCategories/gap/SymmetricCoclosedMonoidalCategoriesDerivedMethods.gi:942]: `RightUnitorWithGivenTensorProduct` expects the third argument to be `a_v x 1` but we have `range = a_v`.

**Review by TKuh — CHANGES_REQUESTED** [review]
In addition to my other remarks, I have the following general questions (for the future) on how to handle situations (I would be especially interested in @zickgraf's opinion):

1. Right now, I noticed, that many of the derivations in `Symmetric*` don't actually need the symmetric structure and could be mathematically placed in `MonoidalClosed` or `MonoidalCoclosed`. What are the reasons that these are in the symmetric files? (E.g. the very first derivation in `SymmetricClosedMonoidalCategoriesDerivedMethods.gi`)
2. One could now also add a `WithGiven` function and derivations for `IsomorphismFromInternalHomToDual` and `IsomorphismFromCoDualToInternalCoHom` (in the rigid case) which would be very similar. Would that be considered nice-to-have, or is it better to only implement things one actually uses? 
3. The derivations in this PR could also additionally have a strict version. Again: nice-to-have or better not? There are also not many derivations which have a strict counterpart, even though they could.

**Comment by zickgraf** [comment]
ad 2.: As far as I know, the primary purpose why all those isomorphisms were added was to derive things in a consistent way e.g. `DualOnObjects` from `InternalHom` etc. (more details next Tuesday). In this setting, having `WithGiven` operations is not useful, because for calling those I must already be able to compute the objects. We have another mechanism for deriving things consistently now (final derivations with exclusion lists), so we could even get rid of the isomorphisms altogether.

ad 3.: Thinking about this, having optimizations for the strict case at all sound dangerous to me, again for reasons of consistency. However, those were added by Sebastian and I'm sure he had this in mind. In any case, the compiler makes these additional derivations obsolete, at least in all categories where composing with a identity cancels on the implementation level.

File: MonoidalCategories/gap/ClosedMonoidalCategories.gd [hunk 7]
```diff
@@ -277,6 +277,18 @@ DeclareOperation( "MorphismFromTensorProductToInternalHomWithGivenObjects",
 DeclareAttribute( "IsomorphismFromDualToInternalHom",
                   IsCapCategoryObject );
 
+#! @Description
+#! The arguments are an object $s = a^{\vee}$,
+#! an object $a$,
+#! and an object $r = \mathrm{\underline{Hom}}(a,1)$.
+#! The argument is an object $a$.
```
[the remark this answers is not in the document]
Comment (mohamed-barakat) [reply MonoidalCategories/gap/ClosedMonoidalCategories.gd:284]: Yes, a copy-paste mistake :)

File: MonoidalCategories/gap/SymmetricClosedMonoidalCategoriesDerivedMethods.gi (hunk 1 above)
Comment (mohamed-barakat) [reply MonoidalCategories/gap/SymmetricClosedMonoidalCategoriesDerivedMethods.gi:914]: Very good, this revealed a mistake :)

File: MonoidalCategories/gap/SymmetricClosedMonoidalCategoriesDerivedMethods.gi (hunk 3 above)
Comment (mohamed-barakat) [reply MonoidalCategories/gap/SymmetricClosedMonoidalCategoriesDerivedMethods.gi:930]: Right.

File: MonoidalCategories/gap/SymmetricCoclosedMonoidalCategoriesDerivedMethods.gi (hunk 6 above)
Comment (mohamed-barakat) [reply MonoidalCategories/gap/SymmetricCoclosedMonoidalCategoriesDerivedMethods.gi:942]: Right.

**Comment by mohamed-barakat** [comment]
> ad 2.: As far as I know, the primary purpose why all those isomorphisms were added was to derive things in a consistent way e.g. `DualOnObjects` from `InternalHom` etc. (more details next Tuesday). In this setting, having `WithGiven` operations is not useful, because for calling those I must already be able to compute the objects. We have another mechanism for deriving things consistently now (final derivations with exclusion lists), so we could even get rid of the isomorphisms altogether.

I agree. And since I don't need this PR anymore, I will close it.

**Pull request closed without merging**